Add possibility to customize default configuration.

dhebbeker opened the pull request:
The user manual (UM2133) suggests that this macro can be configured:
 
> VL53L1_MAX_RANGE_RESULTS [...] **by default**, is equal to four.

(from Rev 6 page 10, highlighting by me)

The value can be passed to the preprocessor as predefined macro.

Using Arduino IDE this can be achieved with help of the [`platform.local.txt`](https://arduino.github.io/arduino-cli/0.20/platform-specification/#platformlocaltxt) configuration file. For example to change the value to `1` add this line to `platform.local.txt`:

    compiler.cpp.extra_flags=-DVL53L1_MAX_RANGE_RESULTS=1

This change applies the same mechanism as already exists for `VL53L1_MAX_USER_ZONES` [some lines further up](https://github.com/dhebbeker/VL53L1/blob/902e9f8788c9d632a1d256e754461434f759c615/src/vl53l1_platform_user_config.h#L95).

dhebbeker committed:
Add #ifndef for all configuration macros.

This way predefined macros may be used to customize the default configuration.

cparata commented:
Hi @dhebbeker ,
I prefer to manage it inside the library without any change in the Arduino IDE. For example you can use the the __has_include directive to override some default values at application level as I did for VL53L5CX on [platform_config.h](https://github.com/stm32duino/VL53L5CX/blob/main/src/platform_config.h) file. In this way, you can add a platform_config_default.h file inside the library to have the default settings for some defines. And then, if you want, you can override the defines at application level adding a platform_config_custom.h like I did in [this](https://github.com/stm32duino/VL53L5CX/tree/main/examples/VL53L5CX_Sat_Multiple_Targets_Per_Zone) example always for VL53L5CX.
Best Regards,
Carlo

dhebbeker commented:
Hello Carlo,

I like your solution which uses __has_include. Unfortunately the compiler (version) I have to use does not support it (only C++11).

But both approaches can be combined. I will adjust my suggestion accordingly.

Cheers
David

dhebbeker committed:
Add possibility to override configuration with custom header file.

dhebbeker commented:
I am using xtensa-lx106-elf-g++. I can not use the newest version (which may support `__has_include`) because of [this bug in the compiler](https://github.com/espressif/esp-idf/issues/4542). That is why I am limited to C++11.

dhebbeker committed:
Formatted using astyle.

Using: https://github.com/stm32duino/actions/blob/b81a4c2b935fb5fa8102e3b9f49288e0dd559a25/astyle-check/scripts/.astylerc

cparata commented on src/vl53l1_platform_user_config.h:
```diff
@@ -30,82 +30,142 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-#define    VL53L1_BYTES_PER_WORD              2
-#define    VL53L1_BYTES_PER_DWORD             4
+#if __cplusplus >= 201703L /* for C++17 or newer */
```
I don't like this check because for example STM32duino Core that uses C++14 supports __has_include directive. It is better to replace this line with:

#if defined __has_include

dhebbeker replied:
Thank you for the tip! This is much better. I had not read that part of the "fine print" of the standard (in C++17 §19.1 ¶5).

I applied the change.

cparata commented on src/vl53l1_platform_user_config.h:
```diff
@@ -30,82 +30,142 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  #if __has_include("platform_config_custom.h")
+    #include "platform_config_custom.h"
+  #endif
+#endif /* end C++17 block */
```
Following the comment above this line becomes:
#endif /* __has_include */

cparata reviewed (changes requested):
Hi @dhebbeker ,
Thanks for the PR. I ask a small modification to do the code more generic.
Best Regards,
Carlo

dhebbeker committed:
Use __has_include instead of __cplusplus in order to decide if __has_include is available.

This is especially useful as some compilers support the __has_include feature independent of the language version.

dhebbeker committed:
Adjusted documentation to requirement for conditional inclusion.

Some compilers may provide conditional inclusion although they do not support C++17.

cparata commented:
Thanks a lot for your contribution David! PR integrated! ;-)

cparata commented:
I have also generated a new release (2.1.0) that includes this new feature

dhebbeker commented:
Thank you Carlo, it is a pleasure to work with you!
